runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcCommitWorkStream.java

```diff
@@ -314,24 +417,35 @@ private Batcher() {
     @Override
     public boolean commitWorkItem(
         String computation, WorkItemCommitRequest commitRequest, Consumer<CommitStatus> onDone) {
-      if (!canAccept(commitRequest.getSerializedSize() + computation.length())) {
+      if (!canAccept(commitRequest.getSerializedSize() + computation.length())
+          || hasReceivedShutdownSignal()) {
         return false;
       }
-      PendingRequest request = new PendingRequest(computation, commitRequest, onDone);
+
+      PendingRequest request = PendingRequest.create(computation, commitRequest, onDone);
       add(idGenerator.incrementAndGet(), request);
       return true;
     }
 
     /** Flushes any pending work items to the wire. */
     @Override
     public void flush() {
-      flushInternal(queue);
-      queuedBytes = 0;
-      queue.clear();
+      try {
+        if (!hasReceivedShutdownSignal()) {
```

Comment:
can we remove? it is handled in stream methods already and doing it here leaves a gap between shutdown check and flush anyway.

Reply:
done